// @kmx/legos/src/utilities/analytics/digitalData.js
import visitorCookie from '../cookies/visitor';
import mykmxCookie from '../cookies/mykmx';

const digitalData = (
    pageInstanceId,
    pageId,
    pageName,
    originPage,
    channel,
    subsection,
    userSegment,
    logOdds,
    akamaiLocation,
) => {
    // handle nulls like a boss
    const location = akamaiLocation || {};

    window.digitalData = {
        pageInstanceID: pageInstanceId,
        page: {
            pageInfo: {
                pageID: pageId,
                pageName,
                originPage,
                channel,
                subsection,
                destinationURL: document.location.toString(),
                referringURL: document.referrer,
            },
        },
        user: [
            {
                attributes: {
                    visitorID: visitorCookie.id || null,
                    isFirstTimeVisitor: visitorCookie.isFirstVisit.toString().toLowerCase(),
                    usingStoreProxy: visitorCookie.isUsingStoreProxy.toString().toLowerCase(),
                    akamaiCity: location.city,
                    akamaiLatitude: location.latitute,
                    akamaiLongitude: location.longitude,
                    akamaiZip: location.zip,
                    kruxID: localStorage ? localStorage.getItem('kxcarmax_kuid') : null,
                    kruxSegments: localStorage ? localStorage.getItem('kxcarmax_allsegs') : null,
                    testSegments: {
                        userInSegment: userSegment.toLowerCase(),
                        logOdds,
                    },
                },
                profile: {
                    profileID: mykmxCookie.id,
                },
            },
        ],
        version: '1.0',
    };
};

export default digitalData;
